src/internal/document_api/mod.rs

```diff
@@ -902,172 +919,164 @@ fn document_create<'a>(
+    let (_, sym_key) = transform::decrypt_plaintext(
+        &recrypt,
+        doc_meta.0.encrypted_symmetric_key.clone().try_into()?,
+        &device_private_key.recrypt_key(),
+    )?;
+
+    aes::decrypt(&mut enc_doc, *sym_key.bytes())
+        .map_err(|e| e.into())
+        .map(move |decrypted_doc| DocumentDecryptResult {
+            id: doc_meta.0.id,
+            name: doc_meta.0.name,
+            created: doc_meta.0.created,
+            updated: doc_meta.0.updated,
+            decrypted_data: decrypted_doc.to_vec(),
         })
 }
 
+
+// TODO reviewers - the decrypt_document_unmanaged code was heavily reworked and should have additional review
+
 /// Decrypt the unmanaged document. The caller must provide both the encrypted data as well as the
 /// encrypted DEKs. Most use cases would want `decrypt_document` instead.
-pub fn decrypt_document_unmanaged<'a, CR: rand::CryptoRng + rand::RngCore>(
-    auth: &'a RequestAuth,
-    recrypt: &'a Recrypt<Sha256, Ed25519, RandomBytes<CR>>,
-    device_private_key: &'a PrivateKey,
-    encrypted_doc: &'a [u8],
-    encrypted_deks: &'a [u8],
-) -> impl Future<Item = DocumentDecryptUnmanagedResult, Error = IronOxideErr> + 'a {
+pub async fn decrypt_document_unmanaged<CR: rand::CryptoRng + rand::RngCore>(
+    auth: &RequestAuth,
+    recrypt: &Recrypt<Sha256, Ed25519, RandomBytes<CR>>,
+    device_private_key: &PrivateKey,
+    encrypted_doc: &[u8],
+    encrypted_deks: &[u8],
+) -> Result<DocumentDecryptUnmanagedResult, IronOxideErr> {
     // attempt to parse the proto as fail-fast validation. If it fails decrypt will fail
-    protobuf::parse_from_bytes::<EncryptedDeksP>(encrypted_deks)
-        .map_err(IronOxideErr::from)
-        .into_future()
-        .join(parse_document_parts(encrypted_doc))
-        .and_then(
-            |(
-                proto_edeks,
-                (
-                    DocumentHeader {
-                        document_id,
-                        segment_id,
-                    },
-                    aes_encrypted_value,
-                ),
-            )| {
-                if document_id.id() == proto_edeks.get_documentId()
-                    && segment_id as i32 == proto_edeks.get_segmentId()
-                {
-                    Ok((document_id, aes_encrypted_value))
-                } else {
-                    Err(IronOxideErr::UnmanagedDecryptionError(
-                        proto_edeks.get_documentId().into(),
-                        proto_edeks.get_segmentId(),
-                        document_id.0,
-                        segment_id as i32,
-                    ))
-                }
-            },
-        )
-        .join(requests::edek_transform::edek_transform(
-            &auth,
-            encrypted_deks,
-        ))
-        .and_then(
-            move |(
-                (document_id, mut enc_data),
-                requests::edek_transform::EdekTransformResponse {
+
+    match try_join!(
+        async {
```

Comment:
Parsing the DEKs, the doc header, and asking the webservice to do the transform are all independent. The two `parse_` functions aren't async and we don't control the protobuf one. In theory this allow all three of these to happen in parallel with the web service call.

Maybe it would be better to combine the two parse functions in one `async` block? Not sure what overhead there is in having the extra one, honestly.

Reply:
Ok. I think it's fine then.